src/main/java/com/google/api/codegen/config/FlatteningConfig.java

```diff
@@ -41,12 +40,37 @@ public static FlatteningConfig createFlattening(
       DiagCollector diagCollector, FlatteningConfigProto flattening, Method method) {
     boolean missing = false;
     ImmutableList.Builder<ImmutableList<Field>> flatteningGroupsBuilder = ImmutableList.builder();
+    Set<Oneof> seenOneofs = new HashSet<>();
     for (FlatteningGroupProto flatteningGroup : flattening.getGroupsList()) {
       ImmutableList.Builder<Field> parametersBuilder = ImmutableList.builder();
       for (String parameter : flatteningGroup.getParametersList()) {
         Field parameterField = method.getInputMessage().lookupField(parameter);
         if (parameterField != null) {
           parametersBuilder.add(parameterField);
+          Oneof oneof = parameterField.getOneof();
+          if (oneof != null) {
+            diagCollector.addDiag(
+                Diag.warning(
+                    SimpleLocation.TOPLEVEL,
+                    "Oneof field flattened: method = %s, message type = %s, oneof = %s, field = %s",
+                    method.getFullName(),
+                    method.getInputMessage().getFullName(),
+                    oneof.getName(),
+                    parameter));
+            if (seenOneofs.contains(oneof)) {
+              diagCollector.addDiag(
+                  Diag.error(
+                      SimpleLocation.TOPLEVEL,
+                      "Multiple oneof fields flattened: method = %s, message type = %s, "
+                          + "oneof = %s, field = %s",
+                      method.getFullName(),
+                      method.getInputMessage().getFullName(),
+                      oneof.getName(),
+                      parameter));
+            } else {
+              seenOneofs.add(oneof);
+            }
+          }
```

Comment:
This is a big chunk of code added to an already large method. It should be broken out into its own package private or protected method (so it can be tested).